src/wallet/wallet.cpp

```diff
@@ -4981,12 +4998,84 @@ void CWallet::GetFilteredNotes(
         filterAddresses.insert(keyIO.DecodePaymentAddress(address));
     }
 
-    GetFilteredNotes(sproutEntries, saplingEntries, filterAddresses, minDepth, INT_MAX, ignoreSpent, requireSpendingKey);
+    GetFilteredNotes(sproutEntries, saplingEntries, filterAddresses, minDepth, INT_MAX, ignoreSpent, requireSpendingKey, true, timestamp);
+}
+
+boost::iterator_range<NotesHelperIndex::by_timestamp_itr> CWallet::GetNotesByType(
+    uint256 paymentaddress_hash,
+    NoteType type,
+    boost::optional<uint64_t> timestamp)
+{
+    int countAlreadySkipped = 0;
+
+    auto instance = NotesHelperIndex::getInstance();
+    auto& index = instance->index.get<by_timestamp>();
+
+    uint64_t ts = 0;
+    if (timestamp) {
+        ts = *timestamp;
+    }
+    auto itr_start = index.upper_bound(std::make_tuple(type, paymentaddress_hash, ts));
+    auto itr_end = index.upper_bound(std::make_tuple(type, paymentaddress_hash));
+
+    return boost::make_iterator_range(itr_start, itr_end);
+}
+
+bool NotesFilters::Common() {
+    if (!CheckFinalTx(*wtx) || wtx->GetDepthInMainChain() < minDepth || wtx->GetDepthInMainChain() > maxDepth)
+        return true;
+
+    if (wtx->IsCoinBase() && wtx->mapSaplingNoteData.empty())
+        return true;
+
+    return false;
+}
+
+bool NotesFilters::Sprout() {
```

Comment:
```suggestion
bool NotesFilters::FilterSprout() {
```
The sense of the return value is reversed from what I would expect.

Reply:
Also, the pattern seems to be to set fields on the filter object and call a method with no arguments, but I would prefer to pass the information about the note to be filtered into the method. The reason for setting fields seems to be to allow some common logic between Sprout and Sapling notes, but that can instead be handled by having `FilterSprout` or `FilterSapling` call that common logic internally; it shouldn't need to be exposed to the client of `NoteFilter`.

Reply:
Done at https://github.com/zcash/zcash/pull/4399/commits/1431921924f32190b4f34f7c8cf1ea81d2957549